definition_validator/validator.cc

```diff
@@ -942,17 +938,68 @@ class ValidateWalk {
             return;
         }
 
+        vector<core::MethodRef> missingMethods;
         for (auto proto : abstract) {
             if (proto.data(gs)->owner == sym) {
                 continue;
             }
 
-            auto mem = sym.data(gs)->findConcreteMethodTransitive(gs, proto.data(gs)->name);
-            if (!mem.exists()) {
-                if (auto e = gs.beginError(loc, core::errors::Resolver::BadAbstractMethod)) {
-                    e.setHeader("Missing definition for abstract method `{}`", proto.show(gs));
-                    e.addErrorLine(proto.data(gs)->loc(), "defined here");
-                    e.addAutocorrect(suggestMethodDefinition(gs, loc, proto));
+            auto concreteMethodRef = sym.data(gs)->findConcreteMethodTransitive(gs, proto.data(gs)->name);
+            if (concreteMethodRef.exists()) {
+                continue;
+            }
+
+            missingMethods.emplace_back(proto);
+        }
+
+        if (!missingMethods.empty()) {
+            if (auto e = gs.beginError(classOrModuleDeclaredAt, core::errors::Resolver::BadAbstractMethod)) {
+                auto pluralization = (missingMethods.size() > 1 ? "s" : "");
+                e.setHeader("Missing definition{} for abstract method{}", pluralization, pluralization);
+
+                auto classOrModuleEndsAt =
+                    core::Loc{classOrModuleDeclaredAt.file(), classDef.loc.copyEndWithZeroLength()};
```

Comment:
`LocOffset` are best understood relative to the file in the enclosing `ctx`. While the way you've written it might work, it's a bit weird to see it written this way. A better way to do this would be to change `validateAbstract` to take a `core::Context ctx` instead of a `core::GlobalState &gs`, and then do this here:

```suggestion
                auto classOrModuleEndsAt = ctx.locAt(classDef.loc.copyEndWithZeroLength());
```

Luckily you are already passing in a `core::Context` at the call sites of this method (it's just that right now, the compiler is implicitly converting them to a `GlobalState &`.

Reply:
Fixed in a000c6c.